// @aspecto/opentelemetry/dist/src/plugins-customizations/orm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sequelizePluginConfig = exports.typeormPluginConfig = void 0;
const recording_span_1 = require("./recording-span");
const opentelemetry_base_1 = require("@aspecto/opentelemetry-base");
const addResponsePayload = (span, response) => {
    const stringified = JSON.stringify(response);
    if (!stringified)
        return;
    const binarySize = Buffer.byteLength(stringified, 'utf8');
    // Limit to 0.5MB.
    if (binarySize > 500000)
        return;
    span.setAttribute(opentelemetry_base_1.AspectoAttributeNames.DB_RESPONSE, stringified);
};
exports.typeormPluginConfig = {
    responseHook: recording_span_1.callCapturePayloadHook(addResponsePayload),
    moduleVersionAttributeName: opentelemetry_base_1.AspectoAttributeNames.MODULE_VERSION,
    collectParameters: true,
};
exports.sequelizePluginConfig = {
    responseHook: recording_span_1.callCapturePayloadHook(addResponsePayload),
    moduleVersionAttributeName: opentelemetry_base_1.AspectoAttributeNames.MODULE_VERSION,
    ignoreOrphanedSpans: true,
};
