// @aspecto/opentelemetry-base/dist/src/attribute-names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessagingExtendedAttribute = exports.LambdaAttributes = exports.AspectoAttributeNames = exports.SetSpansAttributePrefix = exports.ManualSpansAttributePrefix = void 0;
exports.ManualSpansAttributePrefix = 'aspecto.manual';
exports.SetSpansAttributePrefix = 'aspecto.extra';
var AspectoAttributeNames;
(function (AspectoAttributeNames) {
    AspectoAttributeNames["HTTP_REQUEST_BODY"] = "http.request.body";
    AspectoAttributeNames["HTTP_RESPONSE_BODY"] = "http.response.body";
    AspectoAttributeNames["HTTP_PATH"] = "http.path";
    AspectoAttributeNames["MESSAGING_PAYLOAD"] = "messaging.payload";
    AspectoAttributeNames["DB_RESPONSE"] = "db.response";
    AspectoAttributeNames["MODULE_VERSION"] = "instrumented_library.version";
    AspectoAttributeNames["ASPECTO_MANUAL_TYPE"] = "aspecto.manual.type";
    AspectoAttributeNames["ASPECTO_CALLSTACK"] = "aspecto.callstack";
    AspectoAttributeNames["ASPECTO_SAMPLING_RULE_ID"] = "aspecto.sampling.rule.id";
    AspectoAttributeNames["ASPECTO_SAMPLING_LOCAL_SAMPLER_NAME"] = "aspecto.sampling.local_sampler.name";
    AspectoAttributeNames["ASPECTO_SAMPLING_MESSAGING_REMOTE_RULE_ID"] = "aspecto.sampling.messaging_remote.rule.id";
    AspectoAttributeNames["ASPECTO_SAMPLING_UNINITIALIZED"] = "aspecto.sampling.uninitialized";
})(AspectoAttributeNames = exports.AspectoAttributeNames || (exports.AspectoAttributeNames = {}));
var LambdaAttributes;
(function (LambdaAttributes) {
    LambdaAttributes["EVENT"] = "lambda.event";
    LambdaAttributes["RESPONSE"] = "lambda.response";
})(LambdaAttributes = exports.LambdaAttributes || (exports.LambdaAttributes = {}));
var MessagingExtendedAttribute;
(function (MessagingExtendedAttribute) {
    MessagingExtendedAttribute["MESSAGING_PAYLOAD"] = "messaging.payload";
    MessagingExtendedAttribute["MESSAGING_RABBITMQ_CONSUME_END_OPERATION"] = "messaging.rabbitmq.consume_end_operation";
    MessagingExtendedAttribute["MESSAGING_NUMBER_OF_MESSAGES"] = "messaging.number_of_messages";
})(MessagingExtendedAttribute = exports.MessagingExtendedAttribute || (exports.MessagingExtendedAttribute = {}));
