// @aspecto/opentelemetry/dist/src/utils/logging.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.spanLogDetails = exports.SPANS_EXPORT_NS = void 0;
const debug_1 = __importDefault(require("debug"));
exports.SPANS_EXPORT_NS = 'aspecto:spans:export';
function spanLogDetails(span) {
    if (!debug_1.default.enabled(exports.SPANS_EXPORT_NS))
        return null;
    // the default DEBUG_DEPTH=2 and we don't want to have overwhelming output
    return {
        details: {
            span: {
                name: span.name,
                kind: span.kind,
                spanContext: span.spanContext,
                parentSpanId: span.parentSpanId,
                startTime: span.startTime,
                endTime: span.endTime,
                status: span.status,
                attributes: span.attributes,
                links: span.links,
                events: span.events,
                duration: span.duration,
                ended: span.ended,
                resource: span.resource,
                instrumentationLibrary: span.instrumentationLibrary,
            },
        },
    };
}
exports.spanLogDetails = spanLogDetails;
