// @aspecto/opentelemetry/dist/src/aspecto-resource/LabelNames.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GitResourceAttributes = exports.LabelNames = void 0;
var LabelNames;
(function (LabelNames) {
    /** A unique uuid with set for each instance of aspecto tracing collector. */
    LabelNames["HOSTNAME"] = "aspecto.hostname";
    LabelNames["VERSION"] = "aspecto.version";
    LabelNames["TOKEN"] = "aspecto.token";
    LabelNames["RUNTIME_VERSION"] = "aspecto.runtime.version";
    LabelNames["CI_REPORT"] = "aspecto.ci_report";
})(LabelNames = exports.LabelNames || (exports.LabelNames = {}));
/** Not in spec - thus copied here. once added to spec, we can use value from semantic conventions instead */
var GitResourceAttributes;
(function (GitResourceAttributes) {
    GitResourceAttributes["VCS_COMMIT_ID"] = "vcs.commit.id";
    GitResourceAttributes["VCS_CLONE_ID"] = "vcs.clone.id";
    GitResourceAttributes["VCS_BRANCH_NAME"] = "vcs.branch.name";
})(GitResourceAttributes = exports.GitResourceAttributes || (exports.GitResourceAttributes = {}));
