// @aspecto/sampling-rules/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleSource = exports.RulesContextKeys = exports.Operator = exports.Events = void 0;
var Events;
(function (Events) {
    Events["GET_SAMPLING_RULES"] = "get-sampling-rule";
    Events["SAMPLING_RULES"] = "sampling-rules";
})(Events = exports.Events || (exports.Events = {}));
var Operator;
(function (Operator) {
    Operator["EQ"] = "eq";
    Operator["NE"] = "ne";
    Operator["STARTS_WITH"] = "starts_with";
    Operator["ENDS_WITH"] = "ends_with";
    Operator["CONTAINS"] = "contains";
    Operator["NOT_CONTAINS"] = "not_contains";
    Operator["MATCHES"] = "matches";
    Operator["DEFINED"] = "defined";
    Operator["UNDEFINED"] = "undefined";
    Operator["ANY"] = "any";
})(Operator = exports.Operator || (exports.Operator = {}));
// a list of currently supported "features" which a rule can configure on the context
var RulesContextKeys;
(function (RulesContextKeys) {
    RulesContextKeys["MESSAGING_SAMPLING_INHERIT_FROM_PUBLISHER"] = "messaging.sampling.inherit_from_publisher";
})(RulesContextKeys = exports.RulesContextKeys || (exports.RulesContextKeys = {}));
var RuleSource;
(function (RuleSource) {
    RuleSource["MANUAL"] = "manual";
    RuleSource["AUTOMATIC"] = "automatic";
})(RuleSource = exports.RuleSource || (exports.RuleSource = {}));
